jaas/modules/src/main/java/org/apache/karaf/jaas/modules/ldap/LDAPBackingEngine.java

```diff
@@ -0,0 +1,210 @@
+                    String userDNNamespace = result.getNameInNamespace();
+                    // handle case where cn, ou, dc case doesn't match
+                    int indexOfUserBaseDN = userDNNamespace.toLowerCase().indexOf("," + options.getUserBaseDn().toLowerCase());
+                    String userDN = (indexOfUserBaseDN > 0) ?
+                            userDNNamespace.substring(0, indexOfUserBaseDN) :
+                            result.getName();
+
+                    // we need to pull out the cn=, uid=, ect.. from the user name to get the actual user name
+                    String userName = userDN;
+                    if (userDN.contains("=")) userName = userDN.split("=")[1];
+                  
+                    users.add(new UserPrincipal(userName));
+
+
+                }
+            } finally {
+                if (namingEnumeration != null) {
+                    try {
+                        namingEnumeration.close();
+                    } catch (NamingException e) {
+                        // Ignore
+                    }
+                }
+            }
+
+            return users;
+
+        } catch (NamingException e) {
+            throw new RuntimeException(e);
+        }
+
+
+    }
+
+    @Override
+    public List<GroupPrincipal> listGroups(UserPrincipal user) {
+        // for now return empty list, group implementation is not supported
+        return Collections.emptyList();
+    }
+
+    @Override
+    public Map<GroupPrincipal, String> listGroups() {
+        // for now return empty list, group implementation is not supported
+        return Collections.emptyMap();
+    }
+
+    @Override
+    public void addGroup(String username, String group) {
+        throw new RuntimeException("Adding a group is not supporting in LDAP");
+
+    }
+
+    @Override
+    public void createGroup(String group) {
+        throw new RuntimeException("Creating a group is not supporting in LDAP");
+
+    }
+
+    @Override
+    public void deleteGroup(String username, String group) {
+        throw new RuntimeException("Deleting a group is not supporting in LDAP");
+
+    }
+
+    @Override
+    public List<RolePrincipal> listRoles(Principal principal) {
+
+        try {
+            String[] userAndNameSpace = cache.getUserDnAndNamespace(principal.getName());
+            if (userAndNameSpace == null || userAndNameSpace.length < 2) return Collections.emptyList();
+
+            ArrayList<RolePrincipal> roles = new ArrayList<>();
+            for (String role : cache.getUserRoles(principal.getName(), userAndNameSpace[0], userAndNameSpace[1])) {
+                roles.add(new RolePrincipal(role));
+            }
+            return roles;
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    @Override
+    public void addRole(String username, String role) {
+        throw new RuntimeException("Adding a role is not supporting in LDAP");
```

Comment:
What about `UnsupportedOperationException` instead of a generic `RuntimeException` ?